### DISCUSSION TITLE
Use cases and practical limitations of Vercel Edge Functions (vs. Serveless Functions)?

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/community#3123] [Feature Question]
I checked out _[Vercel Edge Explained. Use cases, benefits and limitations](https://upstash.com/blog/vercel-edge)_, but it didn't have too much in regards to practical use cases of edge functions and practical limitations. Looking at the [runtime comparison chart](https://vercel.com/docs/infrastructure/runtime-comparison), already serverless functions seem limited practically (being 50mb max) but edge functions only get ~4mb.

<img width="678" alt="Screenshot 2023-07-04 at 10 57 51 AM" src="https://github.com/vercel/community/assets/28718/2ed83e91-87d6-4385-b0e0-fa2c5330bc77">

I am wondering what are the practical use cases of both serverless functions compared to regular server functions, and edge functions compared to serverless functions. I can see potentially if you have an external postgres database and are making small JSON API requests (< 1mb?) then edge functions seem nice. This might work for example, sending a tweet, or making a facebook post. Editing a document perhaps. But what about file uploads (streaming uploads to S3/Google Cloud)? Does that go above the 1mb-4mb limits? A browser JS file is easily 2MB (I know its less than ideal), but combined with other assets, you could reach 20-50MB for games or other things, server-side rendering, etc.. So it seems edge functions wouldn't be able to work for that kind of stuff, I'm not sure.

Wondering in general what types of things you might want to do with web apps, _which you can't do with either serverless or edge functions._

- Image processing (with tools like Sharp)
- PDF generation
- Web scraping / screenshot generation with puppeteer
- Video/Audio processing / streaming
- Machine learning
- etc.

What is on the top of your mind in regards to what serverless or edge functions are good at (what they are best for), vs. what they can't do? I haven't seen much in terms of what these things _can't_ do. Saying "you want edge functions when you are sending small amounts of data" doesn't paint too vivid of a picture in one's head as to what kinds of things involve sending small amounts of data (other than my tweets example). I understand edge functions are closer to the consumer who makes the request somehow, which is great, but what can I use that for given the limitations on size?

Basically:

- What are the best things you could use **serverless functions** for?
- What are the best things you could use **edge functions** for (as opposed to serverless functions)?
- When is it essential to have a server (and so serverless/edge functions won't work for these cases)? I get that serverless is all about stateless, but can one build any app in a stateless fashion, or are there limitiations?
- What things are _serverless functions_ not good at?
- What things are _edge functions_ not good at, but perhaps serverless functions are?

Thanks so much for your time, I look forward to learning more about this!
